File: 528 random pick with weight/random_pick_weight.py
import random

class random_pick_weight(object):

    def __init__(self, w):
        """
        :type w: List[int]
        """
        self.w = w
        self.n = len(w)
        for i in range(1, self.n):
            w[i] += w[i - 1]
        self.s = w[-1]

    def pickIndex(self):
        """
        :rtype: int
        """
        idx = random.randint(1, self.s)
        left, right = 0, self.n - 1
        while left < right:
            mid = left + (right - left) / 2
            if idx == self.w[mid]:
                return mid
            if idx < self.w[mid]:
                right = mid
            else:
                left = mid + 1
        return left

File: 528 random pick with weight/RandomPickWeight.java
import java.util.Random;

public class RandomPickWeight {
    int[] preSum;
    Random rand;
    public RandomPickWeight(int[] w) {
        preSum = new int[w.length];
        rand = new Random();
        preSum[0] = w[0];
        for (int i = 1; i < w.length; i++) {
            preSum[i] = preSum[i - 1] + w[i];
        }
    }
    
    public int pickIndex() {
        int idx = rand.nextInt(preSum[preSum.length - 1]) + 1;
        int left = 0;
        int right = preSum.length - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (preSum[mid] == idx) return mid;
            if (preSum[mid] < idx) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
